src/DiffSharp.Backends.Reference/Reference.RawTensor.fs

```diff
@@ -0,0 +1,1550 @@
+        (result, t.Shape)
+
+    let inline SoftplusT(t: RawTensorImpl< ^T >) : (^T[] * int[]) =
+        let result = t.Values |> Array.map (fun x -> (max zero< ^T > x) + log(one< ^T > + exp(-abs(x))))
+        (result, t.Shape)
+
+    let inline SigmoidT(t: RawTensorImpl< ^T >) : (^T[] * int[]) =
+        let result = t.Values |> Array.map (fun v -> one / (one + exp -v))
+        (result, t.Shape)
+
+    let inline ExpT(t: RawTensorImpl< ^T >) : (^T[] * int[]) =
+        let result = t.Values |> Array.map exp
+        (result, t.Shape)
+
+    let inline LogT(t: RawTensorImpl< ^T >) : (^T[] * int[]) =
+        let result = t.Values |> Array.map log
+        (result, t.Shape)
+
+    let inline Log10T(t: RawTensorImpl< ^T >) : (^T[] * int[]) =
+        let result = t.Values |> Array.map log10
+        (result, t.Shape)
+        
+    let inline SqrtT(t: RawTensorImpl< ^T >) : (^T[] * int[]) =
+        let result = t.Values |> Array.map sqrt
+        (result, t.Shape)
+        
+    let inline SinT(t: RawTensorImpl< ^T >) : (^T[] * int[]) =
+        let result = t.Values |> Array.map sin
+        (result, t.Shape)
+        
+    let inline CosT(t: RawTensorImpl< ^T >) : (^T[] * int[]) =
+        let result = t.Values |> Array.map cos
+        (result, t.Shape)                
+        
+    let inline TanT(t: RawTensorImpl< ^T >) : (^T[] * int[]) =
+        let result = t.Values |> Array.map tan
+        (result, t.Shape)
+        
+    let inline SinhT(t: RawTensorImpl< ^T >) : (^T[] * int[]) =
+        let result = t.Values |> Array.map sinh
+        (result, t.Shape)
+        
+    let inline CoshT(t: RawTensorImpl< ^T >) : (^T[] * int[]) =
+        let result = t.Values |> Array.map cosh
+        (result, t.Shape)                
+        
+    let inline TanhT(t: RawTensorImpl< ^T >) : (^T[] * int[]) =
+        let result = t.Values |> Array.map tanh
+        (result, t.Shape)
+
+    let inline AsinT(t: RawTensorImpl< ^T >) : (^T[] * int[]) =
+        let result = t.Values |> Array.map asin
+        (result, t.Shape)
+        
+    let inline AcosT(t: RawTensorImpl< ^T >) : (^T[] * int[]) =
+        let result = t.Values |> Array.map acos
+        (result, t.Shape)                
+        
+    let inline AtanT(t: RawTensorImpl< ^T >) : (^T[] * int[]) =
+        let result = t.Values |> Array.map atan
+        (result, t.Shape)
+
+    let inline Random ofDouble (shape:int[]) : (^T[] * int[]) =
+        let values = Array.init (shapeLength shape) (fun _ -> ofDouble (DiffSharp.Util.Random.Uniform()))
+        (values, shape)
+
+    let inline RandomNormal ofDouble (shape:int[]) : (^T[] * int[]) =
+        let values = Array.init (shapeLength shape) (fun _ -> ofDouble (DiffSharp.Util.Random.Normal()))
+        (values, shape)
+
+    let inline RandomIntegers (maxn: int64) ofDouble (shape:int[]) : (^T[] * int[]) =
+        let values = Array.init (shapeLength shape) (fun _ -> ofDouble (float maxn * DiffSharp.Util.Random.Normal()))
```

Comment:
I think this does not implement the correct probabilistic behavior. We can perhaps correct it as:
```fsharp
let values = Array.init (shapeLength shape) (fun _ -> ofDouble (float maxn * DiffSharp.Util.Random.Uniform()))
```
But instead, I think we should be relying on [System.Random.Next](https://docs.microsoft.com/en-us/dotnet/api/system.random.next?view=netcore-3.1#System_Random_Next_System_Int32_System_Int32_) here, adding a little wrapper method called `Integer` to `DiffSharp.Util.Random`.

In any case we need to carefully implement unit tests to ensure the distribution of random integers returned has the correct properties.